### PR TITLE
add checks for group action correctness

### PR DESCRIPTION
**Pull request by kalmarek**
The test examples are based on https://discourse.julialang.org/t/symmetry-sos-optimization-compatibility-issues/98057/1

@blegat this was a major pit(t)a

as bonus I sneaked @compile_workload for 10+ seconds of precompilation ;)

### CODE REVIEW

File: src/group_action_error.jl [hunk 1]
```diff
@@ -0,0 +1,171 @@
+### Basic checks if the action defined by the user is really a group action.
+"""
+    GroupActionError
+`GroupActionError` is an indicator that the group action defined by
+`(G, action, basis)` might be incorrect.
+
+If you receive `GroupActionError` while running functions from `SymbolicWedderburn`
```
Comment (blegat) [inline src/group_action_error.jl:7]: Why would it be advisable if you have already received a `GroupActionError` ?

File: src/group_action_error.jl [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    end
+    return true
+end
+
+function check_group_action(
+    G::Group,
+    hom::InducedActionHomomorphism;
+    full_check = false,
+)
+    if full_check
```
Comment (blegat) [inline src/group_action_error.jl:152]: we could simplify with `n = length(basis(hom)); if !fullcheck; n = min(n, ngens(G))`

File: src/group_action_error.jl [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                    "group identity doesn't act as identity on `x=$(x) ≠ $(xᵉ) = action(act, id, x)`",
+                ),
+            )
+        end
+    end
+
+    for idx in elts_idcs
+        x = basis(ehom)[idx]
+        for g in itr, h in itr
+            passed, xᵍʰ, xᵍ⁾ʰ = __group_action_right_assoc(ehom, g, h, x)
```
Comment (blegat) [inline src/group_action_error.jl:133]:
```suggestion
            passed, xᵍʰ, xʰᵍ = __group_action_right_assoc(ehom, g, h, x)
```

File: src/group_action_error.jl (hunk 1 above)
Comment (kalmarek) [reply src/group_action_error.jl:7]: I meant this as an explanatory docstring for `GroupActionError` (maybe somebody is looking at the source code without encountering it in the wild). What do you suggest.

The other thing I was thinking was to change `check_group_action` into
```julia
    try
        # all the checks
    catch err
        if err isa GroupActionError
            @error "Group action seems to be ill defined; to check the correctness of an action you may run `check_group_action(G, action, basis, full_check=true)`."
        end
        rethrow(err)
    end
```
Then we could remove (or shorten) this docstring.

What do you think?

File: src/group_action_error.jl (hunk 3 above)
Comment (kalmarek) [reply src/group_action_error.jl:133]: it's a distinction between `x^(gh)` and `(x^g)^h`; non ascii version could be `x_gh` and `x_g_h`?

File: src/group_action_error.jl [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    hom::InducedActionHomomorphism;
+    full_check = false,
+)
+    if full_check
+        return __check_group_action_axioms(G, hom, 1:length(basis(hom)))
+    else
+        return __check_group_action_axioms(
+            gens(G),
+            hom,
+            1:min(length(basis(hom)), ngens(G)),
```
[the remark this answers is not in the document]
Comment (kalmarek) [reply src/group_action_error.jl:158]: Indeed that's a very arbitrary, but usually small number ;)

File: src/group_action_error.jl (hunk 1 above)
Comment (blegat) [reply src/group_action_error.jl:7]: What's the advantage over `@error` an rethrowing ?

Comment (blegat) [reply src/group_action_error.jl:7]: > If you receive `GroupActionError` while running functions from `SymbolicWedderburn`, check the error message for a counter example showing why this is not a group action.

> If you encounter inconsistent results it is advisable to run
>     `check_group_action(G, action, basis, full_check=true)`
> to check the correctness of the implementation of the action.

Comment (kalmarek) [reply src/group_action_error.jl]: yeah, I guess there's none.

**Review by blegat — APPROVED** [review]
Maybe reference the GroupsCore tests in the docstring or error

**Comment by kalmarek** [comment]
> Maybe reference the GroupsCore tests in the docstring or error

done!

**Pull request merged**